R/Rules-class.R

```diff
@@ -139,13 +169,101 @@ validObject(.NextBestNCRM())
-{
-    .NextBestNCRM(target=target,
-                  overdose=overdose,
-                  maxOverdoseProb=maxOverdoseProb)
+                         maxOverdoseProb) {
+  .NextBestNCRM(
+    target = target,
+    overdose = overdose,
+    maxOverdoseProb = maxOverdoseProb
+  )
 }
 
+##' Next best dose based on NCRM rule and loss function
+##'
+##' This class is similar to NCRM class, the only difference is the addition of
+##' loss function. As in NCRM rule, first admissible doses are found, which are those
+##' with probability to fall in \code{overdose} category being below
+##' \code{maxOverdoseProb}. Next, within the admissible doses, the loss function is
+##' calculated, i.e. \code{losses} %*% \code{target}. Finally, the corresponding
+##' dose with lowest loss function (Bayes risk) is recommended for the next dose.
+##' @slot target the target toxicity interval (limits included)
+##' @slot overdose the overdose toxicity interval (lower limit excluded, upper limit included)
+##' @slot maxOverdoseProb maximum overdose probability that is allowed
+##' @slot losses loss function
+##'
+##' @example examples/Rules-class-NextBestNCRMLoss.R
+##' @export
+##' @keywords classes
+.NextBestNCRMLoss <-
+  setClass(
+    Class = "NextBestNCRMLoss",
+    representation(
+      target = "numeric",
+      overdose = "numeric",
+      maxOverdoseProb = "numeric",
+      losses = "numeric"
+    ),
+    prototype(
+      target = c(0.2, 0.35),
+      overdose = c(0.35, 0.6, 1),
+      maxOverdoseProb = 0.25,
+      losses = c(1, 0, 1, 2)
+    ),
+    contains = list("NextBest"),
+    validity =
+      function(object) {
+        o <- Validate()
+
+        o$check(
+          is.probRange(object@target),
+          "target has to be a probability range"
+        )
+        o$check(
+          is.probRange(object@overdose[1:2]),
+          "overdose has to be a probability range"
+        )
+        o$check(
+          is.probRange(object@overdose[2:3]),
+          "overdose has to be a probability range"
+        )
+
+        o$check(
+          is.probability(object@maxOverdoseProb),
+          "maxOverdoseProb has to be a probability"
+        )
+
+        o$check(
+          all(!(object@losses < 0)),
+          "losses has to be a vector of non-negative elements"
+        )
+
+        o$result()
+      }
+  )
+validObject(.NextBestNCRMLoss())
+
+##' Initialization function for "NextBestNCRMLoss"
+##'
+##' @param target see \code{\linkS4class{NextBestNCRMLoss}}
+##' @param overdose see \code{\linkS4class{NextBestNCRMLoss}}
+##' @param maxOverdoseProb see \code{\linkS4class{NextBestNCRMLoss}}
+##' @param losses see \code{\linkS4class{NextBestNCRMLoss}}
+##' @return the \code{\linkS4class{NextBestNCRMLoss}} object
+##'
+##' @export
+NextBestNCRMLoss <- function(target,
+                             overdose,
+                             maxOverdoseProb,
+                             losses) {
+  .NextBestNCRMLoss(
```

Comment:
similar here, if you inherit from `NextBestNCRM` directly, then you can first call that constructor and then continue here with the new additions. see e.g. 
https://github.com/Roche/crmPack/blob/9835f38e30a546c28d36438caf49cee251d7e753/R/Model-class.R#L248
for an example how to do this

Reply:
did you try this?

Reply:
If I remember correctly, you said this one is not relevant for my pull review, since I am not adding a lot of new features (or maybe I misunderstood during the meeting)

Reply:
ok